### PR TITLE
Fix #10965: wrap access to sheet.cssRules in try/catch for Firefox SecurityError exception

### LINKED ISSUE #10965: "SecurityError: the operation is insecure." accessing cssRules in Firefox for different domain
The MultiBrowserImpl Live Dev code expects to be able to access the `cssRules` property for stylesheets in the previewed DOM:
* https://github.com/adobe/brackets/blob/b8f92c1b36c2a15396e36b732d7f3fe52e8f4a8d/src/LiveDevelopment/MultiBrowserImpl/protocol/remote/DocumentObserver.js#L60
* https://github.com/adobe/brackets/blob/0ebf3737b8e2fba8c16dc839326338acd0aac2d2/src/LiveDevelopment/MultiBrowserImpl/protocol/remote/LiveDevProtocolRemote.js#L205

In Chrome, a stylesheet from a different domain will not have `cssRules`, so it is safe to test `sheet.cssRules`.  In Firefox, accessing this property will result in a `SecurityError` and break the remote scripts.

Here's a test case that triggers the problem:

```html
<!doctype html>
<html>
  <head>
  <link href="https://stuff.webmaker.org/webmaker-kits/v2/css/style.css" rel="stylesheet" type="text/css">
  <title>SecurityError Bug</title>
  </head>
  <body>
  </body>
</html>
``` 

The fix is to wrap a `try/catch` around all access to `cssRules` so it always returns `undefined` instead of throwing, for example:

```js
var cssRules;
try {
    cssRules = sheet.cssRules;
} catch(e) {
    if(e.name !== "SecurityError") {
        throw e;
    }
    return;
}
```

cc @busykai

### PR DESCRIPTION
**Pull request by humphd**
r? @busykai

### CODE REVIEW

**Comment by humphd** [comment]
This is ready for review, and I'd love to get it landed so I can take it downstream in our fork.

**Comment by busykai** [comment]
1 week overdue on my list. :( will get to it ASAP.

**Comment by js0511** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

oiup2006@naver.com 님께 보내신 메일 <Re: [brackets] Fix #10965: wrap access to sheet.cssRules in try/catch for Firefox SecurityError exception (#10990)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

File: src/LiveDevelopment/MultiBrowserImpl/protocol/remote/DocumentObserver.js [hunk 1]
```diff
@@ -56,17 +56,30 @@
...
-                    rel.stylesheets[sheet.href] = [];
+            var i,
+                href = sheet.href,
+                cssRules;
+
+            // Deal with Firefox's SecurityError when accessing sheets
+            // from other domains. Chrome will safely return `undefined`.
+            try {
+                cssRules = sheet.cssRules;
+            } catch(e) {
```
Comment (busykai) [inline src/LiveDevelopment/MultiBrowserImpl/protocol/remote/DocumentObserver.js]: This does not pass JSLint. Should be `catch (e)`, note the space.

File: src/LiveDevelopment/MultiBrowserImpl/protocol/remote/DocumentObserver.js [hunk 2]
```diff
@@ -56,17 +56,30 @@
...
+                if(e.name !== "SecurityError") {
```
Comment (busykai) [inline src/LiveDevelopment/MultiBrowserImpl/protocol/remote/DocumentObserver.js]: Space is missing after `if` (JSLint).

File: src/LiveDevelopment/MultiBrowserImpl/protocol/remote/LiveDevProtocolRemote.js [hunk 3]
```diff
@@ -202,7 +202,21 @@
                     text = sheet.ownerNode.textContent;
                 } else if (sheet.href === msg.params.url && !sheet.disabled) {
                     var j,
+                        rules;
+
+                    // Deal with Firefox's SecurityError when accessing sheets
+                    // from other domains, and Chrome returning `undefined`.
+                    try {
                         rules = document.styleSheets[i].cssRules;
+                    } catch(e) {
```
Comment (busykai) [inline src/LiveDevelopment/MultiBrowserImpl/protocol/remote/LiveDevProtocolRemote.js]: Space after `catch`.

File: src/LiveDevelopment/MultiBrowserImpl/protocol/remote/LiveDevProtocolRemote.js [hunk 4]
```diff
@@ -202,7 +202,21 @@
...
+                        if(e.name !== "SecurityError") {
```
Comment (busykai) [inline src/LiveDevelopment/MultiBrowserImpl/protocol/remote/LiveDevProtocolRemote.js]: Space after `if`.

File: src/LiveDevelopment/MultiBrowserImpl/protocol/remote/LiveDevProtocolRemote.js [hunk 5]
```diff
@@ -202,7 +202,21 @@
...
+                            throw e;
+                        }
+                    }
+                    if(!rules) {
```
Comment (busykai) [inline src/LiveDevelopment/MultiBrowserImpl/protocol/remote/LiveDevProtocolRemote.js]: Space after `if`.

**Comment by busykai** [comment]
@humphd, this looks good. gave it some testing, works as expected. wanted to add an automated test, but I am not sure if connectivity can actually be assumed, so it's OK.

please correct these formatting comments though. generally, it's a good idea to use Brackets to hack on Brackets. it would run JSLint properly configured for you.

**Comment by humphd** [comment]
I'll fix the JSLint issues, but why does Travis not flag them?  Shouldn't your grunt jshint task reveal this stuff if it's important?

**Comment by busykai** [comment]
@humphd, yes it is a little bit confusing. Travis actually runs `jshint` on brackets sources, but JSLint is still a

**Comment by humphd** [comment]
Updated, thanks for the review.  Hopefully this one can land.

**Comment by busykai** [comment]
Thanks @humphd! And sorry for delay. Merging.

**Pull request merged**